### PR TITLE
Hide Layer & Configure Zoom Level

### LINKED ISSUE #18: Implement Hide Layer
User can hide the layer they don't want to display.

### LINKED ISSUE #20: Configure Zoom Level
- Include Zoom levels as part of Layer settings, similar to name, with default as 0 → 14.

- The layer will be visible only within the configured zoom level

### PR DESCRIPTION
**Pull request by DDG18**
### Problem:
- Users can't hide the specific layer
- Users edit layer by clicking the layer item

### Description:
- Add new option isHide in layerOptions and create hideLayer()
- Implement hide toggle button, once click hide button, layer should be gray out
- implement edit toggle button, once click edit button, display layer configuration panel and gray out other layers when people do updates
### Issues Resolved
#18 

### Image:
https://user-images.githubusercontent.com/47327467/179480521-722301c7-41d4-4385-baa0-d30b4b07656b.mov

### Problem:
- Users can not set layer's zoom level
- Layer is visible all the time

### Description:
- Add minZoom and maxZoom in layerOptions
- Set minZoom default to 0
- Set maxZoom default to 14
- User can slide the range slider to set the zoom levels
- The layer will be visible only within the configured zoom level
 
### Issues Resolved
#20 

### Image:

https://user-images.githubusercontent.com/47327467/179640175-0e82df1b-87d4-4b7a-bffa-4db18942f02f.mov

 
### Check List
- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by junqiu-lei** [comment]
Can we make the zoom number visible when users configure the slide bar?

**Comment by DDG18** [comment]
> Can we make the zoom number visible when users configure the slide bar?

Based on [Elastic UI](https://eui.elastic.co/pr_3350/#/forms/range-sliders), I don't think I can show zoom number as a tooltip above the dragger.

In order to show the zoom level, we can choose prop either  **showInput**
Screen Shot 2022-07-18 at 2 06 32 PM

or **showTicks**
Screen Shot 2022-07-18 at 2 08 28 PM

File: public/common/types/layer.ts [hunk 1]
```diff
@@ -17,6 +17,9 @@ export interface LayerOptions {
   layerType?: LayerTypes;
   typeOptions: LayerTypeOptions;
   isDesatured: boolean;
+  isHide: boolean;
```
Comment (navneet1v) [inline public/common/types/layer.ts:20]: rename this to isHidden.

File: public/components/layer_control/layer_configuration_options.tsx [hunk 2]
```diff
@@ -99,6 +113,28 @@ function LayerConfigurationOptions(props: MapsExlorerOptionsProps) {
...
+        min={0}
+        max={14}
+        step={1}
+        value={[stateParams.layersOptions[configLayerId].minZoom, stateParams.layersOptions[configLayerId].maxZoom]}
+        onChange={setZoom}
+        showLabels
+        showRange
+        // showInput
+        // showTicks
+        // fullWidth
```
Comment (navneet1v) [inline public/components/layer_control/layer_configuration_options.tsx:134]: any reason for these comments?

File: public/components/layer_control/layer_configuration_options.tsx [hunk 3]
```diff
@@ -90,7 +104,7 @@ function LayerConfigurationOptions(props: MapsExlorerOptionsProps) {
 
       <SelectOption
         label={i18n.translate('mapsExplorer.layerVisParams.layerTypeLabel', {
-          defaultMessage: 'Layer Type',
+          defaultMessage: 'Layer Type*',
```
Comment (navneet1v) [inline public/components/layer_control/layer_configuration_options.tsx:107]: why "*" ?

File: public/components/layer_control/layer_collection_panel.tsx [hunk 4]
```diff
@@ -143,24 +155,65 @@ interface LayerCollectionPanelItemProps {
...
+          isEmpty
+          isIconOnly
+          />
+      </EuiFlexItem>
+    </EuiFlexGroup>
+    
+  
+    //   isActive={options.id === configLayerId}
+    //   isDisabled={configLayerId !== undefined && configLayerId !== options.id && isConfigDirty && options.isHide}
+    //   onClick={configLayerId !== undefined && isConfigDirty ? () => { } : () => editLayer(options.id)}
```
Comment (navneet1v) [inline public/components/layer_control/layer_collection_panel.tsx:215]: do we need this commented code?

**Comment by junqiu-lei** [comment]
> > Can we make the zoom number visible when users configure the slide bar?
> 
> Based on [Elastic UI](https://eui.elastic.co/pr_3350/#/forms/range-sliders), I don't think I can show zoom number as a tooltip above the dragger.
> 
> In order to show the zoom level, we can choose prop either **showInput** Screen Shot 2022-07-18 at 2 06 32 PM
> 
> or **showTicks** Screen Shot 2022-07-18 at 2 08 28 PM

Got it, from UX, I think it's better to let users know what exactly zoom number they are configuring on.

**Comment by Shivamdhar** [comment]
> Got it, from UX, I think it's better to let users know what exactly zoom number they are configuring on.

@junqiu-lei are you asking an implementation like this - https://eui.elastic.co/pr_3350/#/forms/range-sliders (Kitchen sink)?

File: public/components/layer_control/layer_configuration_options.tsx [hunk 5]
```diff
@@ -99,6 +113,28 @@ function LayerConfigurationOptions(props: MapsExlorerOptionsProps) {
...
+      <EuiText size="xs">
+         <strong>
+         <FormattedMessage id="mapsExplorer.layerVisParams.zoomLevelLable" defaultMessage="Zoom Level*"/>
+           </strong>
+      </EuiText>
+
+      <EuiDualRange
+        min={0}
+        max={14}
+        step={1}
```
Comment (Shivamdhar) [inline public/components/layer_control/layer_configuration_options.tsx:127]: can we define constants for these numbers?

File: public/common/types/layer.ts (hunk 1 above)
Comment (DDG18) [reply public/common/types/layer.ts:20]: Fixed

File: public/components/layer_control/layer_configuration_options.tsx (hunk 2 above)
Comment (DDG18) [reply public/components/layer_control/layer_configuration_options.tsx:134]: These are the props which can show the zoom value. Junqiu just showed me the latest Kibana UI(attached below), will update soon.
image

File: public/components/layer_control/layer_configuration_options.tsx (hunk 3 above)
Comment (DDG18) [reply public/components/layer_control/layer_configuration_options.tsx:107]: * indicates Required

File: public/components/layer_control/layer_collection_panel.tsx (hunk 4 above)
Comment (DDG18) [reply public/components/layer_control/layer_collection_panel.tsx:215]: Nope. Just deleted.

File: public/components/layer_control/layer_configuration_options.tsx (hunk 5 above)
Comment (DDG18) [reply public/components/layer_control/layer_configuration_options.tsx:127]: Yes, we cab

**Comment by DDG18** [comment]
> > Got it, from UX, I think it's better to let users know what exactly zoom number they are configuring on.
> 
> @junqiu-lei are you asking for an implementation like this - https://eui.elastic.co/pr_3350/#/forms/range-sliders (Kitchen sink)?

That's what we are using right now.

**Comment by junqiu-lei** [comment]
> > Got it, from UX, I think it's better to let users know what exactly zoom number they are configuring on.
> 
> @junqiu-lei are you asking for an implementation like this - https://eui.elastic.co/pr_3350/#/forms/range-sliders (Kitchen sink)?

Yeah, pretty close to this, but it looks like not support show value for dual range.  Looks like Ge is updating the UI to let users able to see the live zoom number.

**Review by junqiu-lei — APPROVED** [review]

File: public/common/types/layer.ts [hunk 6]
```diff
@@ -42,6 +45,9 @@ export const DEFAULT_MAP_EXPLORER_VIS_PARAMS: MapsExplorerVisParams = {
       }),
       layerType: LayerTypes.TMSLayer,
       isDesatured: false,
+      isHidden: false,
+      minZoom: 0,
+      maxZoom: 14,
```
Comment (Shivamdhar) [inline public/common/types/layer.ts:50]: let's replace all these numbers with the same constants across the files : )

File: public/components/layer_control/layer_configuration_options.tsx [hunk 7]
```diff
@@ -42,13 +47,13 @@ export type MapsExlorerOptionsProps = VisOptionsProps<MapsExplorerVisParams> & {
...
-  const [optionValidity, setOptionValidity] = useState(true); 
+  const [optionValidity, setOptionValidity] = useState(true);
 
   //update layer's options
   const setLayerValue = <T extends keyof LayerOptions>(paramName: T, value: LayerOptions[T]) => {
     // during the creatation, when users switch layer type from other types to TMS, 
-    //set optionValidity Trur to ensure users are able to creat new TMS layer
-    if (paramName === "layerType") {setOptionValidity(true);};
+    //set optionValidity True to ensure users are able to creat new TMS layer
+    if (paramName === "layerType") { setOptionValidity(true); };
```
Comment (Shivamdhar) [inline public/components/layer_control/layer_configuration_options.tsx:56]: what all values can paramName have? the check can be reversed and a constant can be used in place of the string.

Comment (DDG18) [reply public/components/layer_control/layer_configuration_options.tsx:56]: - paramName could be any keys in LayerOptions, such as id, name, layerType, typeOptions, minZoom, maxZoon, isHidden..
- Gonna reverse and replace it

File: public/common/types/layer.ts (hunk 6 above)
Comment (DDG18) [reply public/common/types/layer.ts:50]: - Will create the suggested constants file

**Review by Shivamdhar — APPROVED** [review]

**Review by VijayanB — APPROVED** [review]

**Pull request merged**